python/grass/tools/support.py

```diff
@@ -0,0 +1,252 @@
+        """Suggest matching tool names based on provided text.
+
+        The first five tools matching the condition are returned,
+        so this always returns at most five names, but it may not be the best matches.
+        The returned names are sorted alphabetically (not by priority).
+        This specific behavior may change in the future versions.
+        """
+        result = []
+        max_suggestions = 5
+        for name in self.names():
+            if ToolFunctionResolver.levenshtein_distance(text, name) < len(text) / 2:
+                result.append(name)
+            if len(result) >= max_suggestions:
+                break
+        result.sort()
+        return result
+
+    def names(self):
+        if self._names:
+            return self._names
+        self._names = [name.replace(".", "_") for name in gs.get_commands()[0]]
+        return self._names
+
+
+class ToolResult:
+    """Result returned after executing a tool"""
+
+    def __init__(self, *, name, command, kwargs, returncode, stdout, stderr):
+        self._name = name
+        self._command = command
+        self._kwargs = kwargs
+        self.returncode = returncode
+        self._stdout = stdout
+        self._stderr = stderr
+        self._text = None
+        self._cached_json = None
+
+    @property
+    def text(self) -> str | None:
+        """Text output as decoded string"""
+        if self._text is not None:
+            return self._text
+        if self._stdout is None:
+            return None
+        if isinstance(self._stdout, bytes):
+            decoded_stdout = gs.decode(self._stdout)
+        else:
+            decoded_stdout = self._stdout
+        self._text = decoded_stdout.strip()
+        return self._text
+
+    @property
+    def keyval(self) -> dict:
+        """Text output read as key-value pairs separated by equal signs"""
+
+        def conversion(value):
+            """Convert text to int or float if possible, otherwise return it as is"""
+            try:
+                return int(value)
+            except ValueError:
+                pass
+            try:
+                return float(value)
+            except ValueError:
+                pass
+            return value
+
+        return gs.parse_key_val(self._stdout, val_type=conversion)
+
+    @property
+    def comma_items(self) -> list:
+        """Text output read as comma-separated list"""
+        return self.text_split(",")
+
+    @property
+    def space_items(self) -> list:
+        """Text output read as whitespace-separated list"""
+        return self.text_split(None)
+
+    def text_split(self, separator=None) -> list:
+        """Parse text output read as list separated by separators
+
+        Any leading or trailing newlines are removed prior to parsing.
+        """
+        # The use of strip is assuming that the output is one line which
+        # ends with a newline character which is for display only.
+        return self.text.split(separator)
```

Comment:
Pylance and Mypy complains because above, `self.text` is annotated with `str | None`. If `self.text` can be none, it will cause an exception.

A test of this function with `self.text = None` could be useful, and would specify the wanted handling when this occurs